### PR TITLE
Adds block property to socket io results

### PR DESCRIPTION
**Pull request by davinci26**
Signed-off-by: Sotiris Nanopoulos <sonanopo@microsoft.com>

In preparation for #13189 we add the following functionality to the internal API:
1. `getEnabled ` to `file_event.h` which returns the active events in a file event. This is useful to 
2. `didBlock` to `IoResult` to know if the last I/O call blocked.

Risk Level: Low
Testing: Tests
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

File: include/envoy/network/transport_socket.h [hunk 1]
```diff
@@ -33,6 +33,11 @@ struct IoResult {
    * can only be true for read operations.
    */
   bool end_stream_read_;
+
+  /**
+   * True if the I/O call blocked.
+   */
+  bool didBlock_;
```
Comment (nigriMSFT) [inline include/envoy/network/transport_socket.h:40]: I think would_block_ describes this more accurately. All of the IO calls should be non-blocking and will return EWOULDBLOCK if the operation were to block. So no blocking should actually occur

Comment (davinci26) [reply include/envoy/network/transport_socket.h:40]: Good point, will change

File: source/common/network/raw_buffer_socket.cc [hunk 2]
```diff
@@ -16,6 +16,7 @@ void RawBufferSocket::setTransportSocketCallbacks(TransportSocketCallbacks& call
 IoResult RawBufferSocket::doRead(Buffer::Instance& buffer) {
   PostIoAction action = PostIoAction::KeepOpen;
   uint64_t bytes_read = 0;
+  bool didBlock = true;
```
Comment (nigriMSFT) [inline source/common/network/raw_buffer_socket.cc:19]: Is this supposed to be initialized to false and set to true when (result.err_->getErrorCode() == Api::IoError::IoErrorCode::Again)

File: source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc [hunk 3]
```diff
@@ -92,12 +94,13 @@ Network::IoResult UpstreamProxyProtocolSocket::writeHeader() {
                      result.err_->getErrorDetails());
       if (result.err_->getErrorCode() != Api::IoError::IoErrorCode::Again) {
         action = Network::PostIoAction::Close;
+        didBlock = true;
```
Comment (nigriMSFT) [inline source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc:97]: Another !=/== discrepancy with Api::IoError::IoErrorCode::Again

**Comment by nigriMSFT** [comment]
Do we have a test for the would_block status flag? Something analogous to TEST_P(UdpListenerImplTest, SendDataError) where we request IO, the underlying syscall returns EWOULDBLOCK, and we verify that the IoResult has the would_block flag set.

**Comment by davinci26** [comment]
> Do we have a test for the would_block status flag? Something analogous to TEST_P(UdpListenerImplTest, SendDataError) where we request IO, the underlying syscall returns EWOULDBLOCK, and we verify that the IoResult has the would_block flag set.

Agreed will add some tests to verify the property added

File: include/envoy/network/transport_socket.h (hunk 1 above)
Comment (davinci26) [reply include/envoy/network/transport_socket.h]: fixed in f36dad1d76eb51ef5563089ab14de1ff9eae8800

File: source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc (hunk 3 above)
Comment (davinci26) [reply source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc]: fixed in c7b20241c5eeed82d8d8ed356a8ac79d2d935c9c

File: source/common/network/raw_buffer_socket.cc (hunk 2 above)
Comment (davinci26) [reply source/common/network/raw_buffer_socket.cc]: fixed in c7b20241c5eeed82d8d8ed356a8ac79d2d935c9c

File: include/envoy/network/transport_socket.h [hunk 4]
```diff
@@ -33,6 +33,11 @@ struct IoResult {
    * can only be true for read operations.
    */
   bool end_stream_read_;
+
+  /**
+   * True if the I/O call would block.
+   */
+  bool would_block;
```
Comment (nigriMSFT) [inline include/envoy/network/transport_socket.h:40]: Would the convention be would_block_?

File: source/common/network/raw_buffer_socket.cc [hunk 5]
```diff
@@ -16,6 +16,7 @@ void RawBufferSocket::setTransportSocketCallbacks(TransportSocketCallbacks& call
 IoResult RawBufferSocket::doRead(Buffer::Instance& buffer) {
   PostIoAction action = PostIoAction::KeepOpen;
   uint64_t bytes_read = 0;
+  bool wouldBlock = false;
```
Comment (nigriMSFT) [inline source/common/network/raw_buffer_socket.cc:19]: Would the convention be would_block?

File: source/common/network/raw_buffer_socket.cc [hunk 6]
```diff
@@ -38,18 +39,20 @@ IoResult RawBufferSocket::doRead(Buffer::Instance& buffer) {
       ENVOY_CONN_LOG(trace, "read error: {}", callbacks_->connection(),
                      result.err_->getErrorDetails());
       if (result.err_->getErrorCode() != Api::IoError::IoErrorCode::Again) {
+        wouldBlock = true;
```
Comment (nigriMSFT) [inline source/common/network/raw_buffer_socket.cc:42]: I think this is still wrong. Shouldn't we set would_block to true if (result.err_->getErrorCode() == Api::IoError::IoErrorCode::Again)

File: source/common/network/raw_buffer_socket.cc (hunk 5 above)
Comment (davinci26) [reply source/common/network/raw_buffer_socket.cc:19]: ack, I think you are right again 😄

File: source/common/network/raw_buffer_socket.cc (hunk 6 above)
Comment (davinci26) [reply source/common/network/raw_buffer_socket.cc:42]: I am trying to set up tests for all the sockets so we can make sure that we got all of them correctly

**Comment by davinci26** [comment]
Based on the discussion on slack channel https://envoyproxy.slack.com/archives/C78HA81DH/p1602261758081000
fd and file_events_ have a 1-1 to mapping as a result this PR is not the best way to go forward.

Closing in favour of https://github.com/envoyproxy/envoy/pull/13478

**Pull request closed without merging**